### PR TITLE
Polygon Connector

### PR DESCRIPTION
**Pull request by liu-zhipeng**
## Type of change

- [ ] Docs change / dependency upgrade
- [ ] Configuration / tooling changes
- [ ] Refactoring
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires changes in customer code

## Related Issue(s)

Fixes 
https://github.com/connext/nxtp/issues/1727

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,7 +32,7 @@
     "packages/examples/bridge-reference"
   ],
   "scripts": {
-    "lint:all": "yarn workspaces foreach --exclude @connext/nxtp-subgraph --exclude @connext/nxtp-integration run lint --fix",
+    "lint:all": "NODE_OPTIONS=--max_old_space_size=4096 yarn workspaces foreach --exclude @connext/nxtp-subgraph --exclude @connext/nxtp-integration run lint --fix",
```
Comment (LayneHaber) [inline package.json:35]: lets add this back in only if needed

File: packages/deployments/contracts/contracts/core/messaging/connectors/OptimismConnector.sol [hunk 2]
```diff
@@ -145,5 +147,8 @@ contract OptimismL1Connector is BaseOptimismConnector {
     require(_data.length == 32, "!length");
     // set the outbound root for optimism
     IRootManager(ROOT_MANAGER).setOutboundRoot(mirrorDomain, bytes32(_data));
+    // get the state commitment root
+    // if state commitment root is <
+    emit MessageProcessed(_data, msg.sender);
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/connectors/OptimismConnector.sol:152]: we can move this to the top level (that's where they are moved to on main)

File: packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    address _rootManager,
+    address _mirrorConnector,
+    uint256 _mirrorProcessGas,
+    uint256 _processGas,
+    uint256 _reserveGas
+  )
+    Connector(_domain, _mirrorDomain, _amb, _rootManager, _mirrorConnector, _mirrorProcessGas, _processGas, _reserveGas)
+    FxBaseChildTunnel(_amb)
+  {
+    // setFxRootTunnel(_mirrorConnector);
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol:32]: do we need this still?

File: packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  // ============ Private fns ============
+
+  function _verifySender(address _expected) internal view override returns (bool) {
+    require(msg.sender == AMB, "!bridge");
+    return true;
+  }
+
+  function _sendMessage(bytes memory _data) internal override {
+    // Simply emit `MessageSent(_data)`
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol:43]: lets delete this comment

File: packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  constructor(
+    uint32 _domain,
+    uint32 _mirrorDomain,
+    address _amb,
+    address _rootManager,
+    address _mirrorConnector,
+    uint256 _mirrorProcessGas,
+    uint256 _processGas,
+    uint256 _reserveGas,
+    address _checkPointManager
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol:73]: where is the `_checkPointManager` used?

File: packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+// address constant MUMBAI_FX_CHILD = 0xCf73231F28B7331BBe3124B907840A94851f9f11;
+// address constant GOERLI_CHECKPOINT_MANAGER = 0x2890bA17EfE978480615e330ecB65333b880928e;
+// address constant GOERLI_FX_ROOT = 0x3d1d3E34f7fB6D26245E6640E1c50710eFFf15bA;
+
+// address constant MAINNET_FX_CHILD = 0x8397259c983751DAf40400790063935a11afa28a;
+// address constant MAINNET_CHECKPOINT_MANAGER = 0x86E4Dc95c7FBdBf52e33D563BbDB00823894C287;
+// address constant MAINNET_FX_ROOT = 0xfe5e5D361b2ad62c541bAb87C45a0B9B018389a2;
+
+contract PolygonL2Connector is Connector, FxBaseChildTunnel {
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol:17]: make sure to check for the open zeppelin contracts here in the crosschain library, see [here](https://docs.openzeppelin.com/contracts/4.x/api/crosschain#CrossChainEnabledPolygonChild) and in the GitHub

Comment (LayneHaber) [reply packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol:17]: im not sure if you can use them directly but if possible we should try to

File: packages/deployments/contracts/contracts/core/messaging/helpers/polygon/lib/Merkle.sol [hunk 7]
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: MIT
+// @unsupported: ovm
+pragma solidity 0.8.15;
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/core/messaging/helpers/polygon/lib/Merkle.sol:3]: we should link the source to all of these library contracts (also, if we can import them from a package that would be even better)

File: packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol (hunk 5 above)
Comment (liu-zhipeng) [reply packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol:72]: `_checkPointManager` is used for construction arg of `FxBaseRootTunnel`.
https://github.com/connext/nxtp/blob/c5e3b40039a5e3a508687669e97d4356651c96de/packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol#L75

File: packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol (hunk 6 above)
Comment (liu-zhipeng) [reply packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol:17]: `CrossChainEnabledPolygonChild` of the openzeppelin is only for L2 chain(mumbai or polygon). we can't use that for now.

File: packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol (hunk 3 above)
Comment (liu-zhipeng) [reply packages/deployments/contracts/contracts/core/messaging/connectors/PolygonConnector.sol]: we don't need when deploying, but we should have extra tx for set fxRootTunnel and fxChildTunnel while configuring mirror connectors.

**Review by LayneHaber — APPROVED** [review]

**Comment by LayneHaber** [comment]
we really should have unit tests in here, but since this was tested against testnet we can take on this as tech debt (same state as Optimism, Gnosis, and Mainnet connectors)

**Pull request merged**